File: Python/algorithmic_heights/06_maj.py
# Majority Element
import sys

from collections import Counter


def convert_to_intlist(line):
    return tuple(int(item) for item in line.split())


def read_arrays(reader, k):
    return [convert_to_intlist(next(reader)) for _ in range(k)]


def calc_majority_elements(arrays, n):
    def majority_element(array):
        if array:
            counter = Counter(array)
            most_common, count = counter.most_common(n=1)[0]
            if count > n / 2:
                return most_common
        return -1

    return [majority_element(array) for array in arrays]


def main():
    reader = sys.stdin
    k, n = convert_to_intlist(next(reader))
    arrays = read_arrays(reader, k)
    result = calc_majority_elements(arrays, n)
    print(' '.join(map(str, result)))


if __name__ == '__main__':
    main()

File: Haskell/algorithmic_heights/06_maj.hs
{-# LANGUAGE TupleSections #-}

import Control.Monad (replicateM)
import Data.IntMap (IntMap, fromListWith)
import qualified Data.IntMap as M

convertToIntList :: String -> [Int]
convertToIntList = map read . words


readParameters :: String -> (Int, Int)
readParameters line =
    case convertToIntList line of
        [a, b] -> (a, b)
        _ -> error "Malformed input."


readLists :: Int -> IO [[Int]]
readLists k = do
    rows <- replicateM k getLine
    return $ map convertToIntList rows


calcMajorityElements :: [[Int]] -> Int -> [Int]
calcMajorityElements lists n =
    let majorityelement xs =
            let occurrences = fromListWith (+) $ map (, 1) xs
                findMax item cnt (mostCommon, maxCount) =
                    if cnt > maxCount then (item, cnt)
                    else (mostCommon, maxCount)
                (mostCommon, count) = M.foldrWithKey findMax (0, 0) occurrences
            in if count > n `div` 2 then mostCommon else -1
    in map majorityelement lists


main :: IO ()
main = do
    (k, n) <- readParameters <$> getLine
    lists <- readLists k
    let result = calcMajorityElements lists n
    putStrLn $ unwords $ map show result
